lms/djangoapps/mobile_api/tests.py

```diff
@@ -21,35 +26,233 @@
 )
 
 
-@ddt.ddt
-class TestMobileApiUtils(ModuleStoreTestCase, APITestCase):
+class MobileAPITestCase(ModuleStoreTestCase, APITestCase):
     """
-    Tests for mobile API utilities
+    Base class for testing Mobile APIs.
+    Subclasses are expected to define REVERSE_NAME to be used for django reverse URL.
+    They may also override any of the methods defined in this class to control the behavior of the TestMixins.
     """
-
     def setUp(self):
+        super(MobileAPITestCase, self).setUp()
+        self.course = CourseFactory.create(mobile_available=True)
         self.user = UserFactory.create()
+        self.password = 'test'
+        self.username = self.user.username
+
+    def tearDown(self):
+        super(MobileAPITestCase, self).tearDown()
+        self.logout()
+
+    def login(self):
+        """Login test user."""
+        self.client.login(username=self.username, password=self.password)
+
+    def logout(self):
+        """Login test user."""
+        self.client.logout()
+
+    def enroll(self, course_id=None):
+        """Enroll test user in test course."""
+        CourseEnrollment.enroll(self.user, course_id or self.course.id)
+
+    def unenroll(self, course_id=None):
+        """Unenroll test user in test course."""
+        CourseEnrollment.unenroll(self.user, course_id or self.course.id)
+
+    def login_and_enroll(self, course_id=None):
+        """Shortcut for both login and enrollment of the user."""
+        self.login()
+        self.enroll(course_id)
+
+    def get_response(self, reverse_args=None, expected_response_code=200, **kwargs):
+        """
+        Helper method for calling endpoint, verifying and returning response.
+        If expected_response_code is None, doesn't verify the response' status_code.
+        """
+        url = self.get_url(reverse_args, **kwargs)
+        response = self.url_method(url, **kwargs)
+        if expected_response_code is not None:
+            self.assertEqual(response.status_code, expected_response_code)
+        return response
+
+    def get_url(self, reverse_args=None, **kwargs):  # pylint: disable=unused-argument
+        """Base implementation that returns URL for endpoint that's being tested."""
+        return reverse(self.REVERSE_NAME, kwargs=reverse_args)  # pylint: disable=no-member
+
+    def url_method(self, url, **kwargs):  # pylint: disable=unused-argument
+        """Base implementation that returns response from the GET method of the URL."""
+        return self.client.get(url)
+
+
+class MobileCourseAPITestCase(MobileAPITestCase):
```

Comment:
I was thinking afterwards I should have added further documentation/description about the class and mixin hierarchy.

First, the rationale for test class versus test mixin:
1. To prevent the same test method from being run multiple times for a given test class, all inherited test methods are defined in mixins - which generally do not have a hierarchy - and are explicitly included by concrete test classes.   (Alternatively, if a test was implemented in a common base class inherited multiple times by a test class, it would be run multiple times).

2. To keep the mixin usage simpler (and avoid multiple inheritance issues), each concrete test class simply includes the appropriate test mixin that applies to its endpoint.

So here's the resulting class and mixin hierarchy:
MobileAPITestCase
  <-- MobileCourseAPITestCase
  <-- MobileUserAPITestCase
  <-- MobileUserCourseAPITestCase

MobileAuthTestMixin *(is_user=False)*
  <-- MobileAuthUserTestMixin
MobileCourseAccessTestMixin *(verify_enrolled=False)*
  <-- MobileEnrolledCourseAccessTestMixin